pkg/ccl/changefeedccl/sink_pubsub.go

```diff
@@ -0,0 +1,526 @@
+	p.flushDone = make(chan struct{}, 1)
+
+	for i := 0; i < p.numWorkers; i++ {
+		//initialize worker goroutine and channel for worker
+		p.eventsChans[i] = make(chan pubsubMessage)
+		j := i
+		p.workerGroup.GoCtx(func(ctx context.Context) error {
+			p.workerLoop(j)
+			return nil
+		})
+	}
+}
+
+// workerLoop consumes any message sent to the channel corresponding to the worker index
+func (p *pubsubSink) workerLoop(workerIndex int) {
+	for {
+		select {
+		case <-p.workerCtx.Done():
+			return
+		case msg := <-p.eventsChans[workerIndex]:
+			if msg.isFlush {
+				// Signals a flush request, makes sure that the messages in eventsChans are finished sending
+				continue
+			}
+			m := msg.message
+			b, err := json.Marshal(m)
+			if err != nil {
+				p.exitWorkersWithError(err)
+			}
+			err = p.client.sendMessage(b, msg.topicID, string(msg.message.Key))
+			if err != nil {
+				p.exitWorkersWithError(err)
+			}
+			msg.alloc.Release(p.workerCtx)
+		}
+	}
+}
+
+// exitWorkersWithError sends an error to the sink error channel
+func (p *pubsubSink) exitWorkersWithError(err error) {
+	// errChan has buffer size 1, first error will be saved to the buffer and
+	// subsequent errors will be ignored
+	select {
+	case p.errChan <- err:
+		p.exitWorkers()
+	default:
+	}
+}
+
+// sinkError checks if there is an error in the error channel
+func (p *pubsubSink) sinkError() error {
+	select {
+	case err := <-p.errChan:
+		return err
+	default:
+	}
+	return nil
+}
+
+// workerIndex hashes key to return a worker index
+func (p *pubsubSink) workerIndex(key []byte) uint32 {
+	return crc32.ChecksumIEEE(key) % uint32(p.numWorkers)
+}
+
+// flushWorkers sends a flush message to every worker channel and then signals sink that flush is done
+func (p *pubsubSink) flushWorkers() error {
+	for i := 0; i < p.numWorkers; i++ {
+		//flush message will be blocked until all the messages in the channel are processed
+		select {
+		case <-p.workerCtx.Done():
+			return p.workerCtx.Err()
+		case p.eventsChans[i] <- pubsubMessage{isFlush: true}:
+		}
+	}
+
+	// flush messages within topic
+	p.client.flushTopics()
+
+	select {
+	// signals sink that flush is complete
+	case <-p.workerCtx.Done():
+		return p.workerCtx.Err()
+	case p.flushDone <- struct{}{}:
+		return nil
+	}
+}
+
+// Dial connects to gcp client and opens a topic
+func (p *gcpPubsubClient) openTopics(withTopicName string) error {
+	var client *pubsub.Client
+	var err error
+
+	creds, err := getGCPCredentials(p.ctx, p.url)
+	if err != nil {
+		return err
+	}
+	// Sending messages to the same region ensures they are received in order
+	// even when multiple publishers are used.
+	// region can be changed from query parameter to config option
+	if p.region != "" {
+		client, err = pubsub.NewClient(p.ctx, p.projectID, option.WithCredentials(creds), option.WithEndpoint(p.region))
+	} else {
+		client, err = pubsub.NewClient(p.ctx, p.projectID, option.WithCredentials(creds), option.WithEndpoint(p.region))
```

Comment:
I don't see the difference with the above branch.  Seems like we still use WithEndpoint option even if region is empty?